Debugging API changes in Mesos

adam-mesos opened the pull request:
This change bumps Mesos (and mesos-modules to match) to include debugging fixes.
Also bumping AdminRouter to pickup some debugging support.
But most significant is a dcos-integration test of the new Mesos Debug API calls.

# Issues

[DCOS-9574](https://mesosphere.atlassian.net/browse/DCOS-9574)

# Checklist

 - [x] Included a test which will fail if code is reverted but test is not
 - [X] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

# Component updates:
Mesos

 - [X] Change Log from last: 
https://github.com/mesosphere/mesos/compare/dcos-mesos-master-30d6edb...mesosphere:dcos-mesos-master-e99ea9c
(mostly tests/fixes for debugging)
 - [x] Test Results: https://mesosphere.slack.com/archives/core-ci/p1481694252000143 
(3 known cmake build errors, 1 consistently failing test soon-to-be-fixed, and several known flaky tests)
 - [ ] Code Coverage: N/A

Mesos-modules

 - [X] Change Log from last: https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae 
(fix build for Mesos interface changes; log-sink module for #1000 and upgrade-fix slipped in too)
 - [ ] Test Results: N/A
 - [ ] Code Coverage: N/A

Adminrouter

 - [X] Change Log from last: https://github.com/dcos/adminrouter/compare/ae8d4d8328599d67e7ca8d9246229ff1420e792c...d44fd04d38db485145dd8c0fd839cc44c648c675 
("Disable buffering to mesos agent API and use HTTP/1.1" for debugging support)
 - [ ] Test Results: N/A
 - [ ] Code Coverage: N/A

adam-mesos commented:
I've fixed all my TODOs and the CI is green, so I'd say this is ready for review. Maybe @mellenburg and @klueska can tell me how bad my python is.

mellenburg commented on test_util/recordio.py:
```diff
@@ -0,0 +1,155 @@
+
+                try:
+                    self.length = int(self.buffer.decode("UTF-8"))
+                except Exception as exception:
```
As this commit is written now, this try/catch/raise DCOSException is never really utilized so when this does get hit, everything will come to a halt anyway, thus, just leave it out.

However, if you envision this being used elsewhere, I could get the idea of a failed state/custom exception. In that case, just make in this module:
```
class RecordIoDecoderException(Exception):
    pass
```
or call it RecordIoCodecError if you want to use one error that also covers the encoder. In either case, those names will be more usable than "DCOSExpcetion" when this has nothing to do with DC/OS on the surface

adam-mesos replied:
Dropped it in favor of `Exception`

mellenburg commented on test_util/recordio.py:
```diff
@@ -0,0 +1,155 @@
+        return bytes(str(len(s)) + "\n", "UTF-8") + s
+
+
+class Decoder(object):
```
As above, no need to inherit from object in python3

mellenburg commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -0,0 +1,140 @@
+    def _input_streamer(encoder, nested_container_id, input_data):
+        message = {'type': 'ATTACH_CONTAINER_INPUT', 'attach_container_input': {'type': 'CONTAINER_ID'}}
+        message['attach_container_input']['container_id'] = nested_container_id
+        yield encoder.encode(message)
```
Does this need to be a generator or a function at all? Could we not just construct this payload before the post? If this must be a method, then I would remove the input_data and encoder arguments and just hardcode them within this method as they are only called once in a totally static manner

adam-mesos replied:
Yeah, I just need an iterable, so an array should work

adam-mesos replied:
Hmm, didn't work on first attempt, but maybe you have some tips now that the rest of it's cleaned up.

adam-mesos commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -0,0 +1,140 @@
+    # Launch a basic marathon app (no image), so we can debug into it!
+    app, test_uuid = get_test_app()
+    test_app_id = 'integration-test-{}'.format(test_uuid)
+    cluster.marathon.deploy_app(app)
```
Unfortunately, `deploy_and_cleanup` destroys the app before I can get a hold of the container to debug into it. I need something that only cleans up after I'm done with it, since I need a live running task for the duration of my test. Something like a test Teardown?

adam-mesos commented on test_util/errors.py:
```diff
@@ -0,0 +1,91 @@
+import abc
```
Agreed. I just copied the recordio file from dcos-cli and didn't bother to strip out the unnecessary errors.py dependency. It's gone now.

adam-mesos commented:
Rebased against dcos/dcos master, bumped to latest Mesos master (and modules to match), and cleaned up the integration test according to @mellenburg 's feedback. Two things I wasn't able to address:
1. My first attempt at inlining `input_streamer` into an iterable array failed, so I just pulled it up into a single-parameter top-level function. Maybe python streaming requests (not responses) expects a generator?
2. I can't use `marathon.deploy_and_cleanup` because it destroys the app as soon as it verifies that it's healthy, which means that by the time I query for running tasks/containers to debug into, the app+task has already been killed.

Ready for review again!

mellenburg commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -0,0 +1,135 @@
+            if app_id in task['id']:
+                container_id = task['statuses'][0]['container_status']['container_id']['value']
+                agent_id = task['slave_id']
+    if container_id is None:
```
All of these
```
if foo is None:
    raise Exception(my_message)
```
are suboptimal. When using pytest, always try to use plain asserts. The following is identical:
```
assert foo is None, my_message
```

mellenburg commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -0,0 +1,135 @@
+                agent_id = task['slave_id']
+    if container_id is None:
+        raise Exception('Container ID not found for instance of app_id {}'.format(app_id))
+    if agent_id is None:
```
See above on assert pattern

mellenburg commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -0,0 +1,135 @@
+    for agent in state['slaves']:
+        if agent['id'] == agent_id:
+            agent_hostname = agent['hostname']
+    if agent_hostname is None:
```
See above on assert pattern

mellenburg commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -0,0 +1,135 @@
+    decoder = Decoder(lambda s: json.loads(s.decode("UTF-8")))
+    for chunk in launch_output.iter_content():
+        for r in decoder.decode(chunk):
+            if r['type'] == 'DATA':
```
This test can pass if there are no records that are of the data type. Should we add a check so that this looper will fail if this if statement is never triggered?

mellenburg commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -0,0 +1,135 @@
+    # Verify the message from the attached output stream
+    for chunk in attached_output.iter_content():
+        for r in decoder.decode(chunk):
+            if r['type'] == 'DATA':
```
See above on if we need an assertion to check if this if block is never hit

alberts commented on packages/adminrouter/buildinfo.json:
```diff
@@ -9,7 +9,7 @@
           "kind": "git",
           "git": "https://github.com/dcos/adminrouter.git",
-          "ref": "ae8d4d8328599d67e7ca8d9246229ff1420e792c",
+          "ref": "d44fd04d38db485145dd8c0fd839cc44c648c675",
```
rebase and drop this change -- adminrouter has been bumped to latest already

branden commented on test_util/recordio.py:
```diff
@@ -0,0 +1,150 @@
+                self.state = self.RECORD
+
+                # Note that for 0 length records, we immediately decode.
+                if self.length <= 0:
```
Let's assert length is nonnegative after we decode it, to fail earlier on bad encodings and simplify length comparisons.

adam-mesos commented on packages/dcos-integration-test/extra/test_mesos.py:
```diff
@@ -0,0 +1,135 @@
+    # Cannot use deploy_and_cleanup because we must attach to a running app/task/container.
+    app, test_uuid = get_test_app()
+    app_id = 'integration-test-{}'.format(test_uuid)
+    cluster.marathon.deploy_app(app)
```
Magic.

adam-mesos commented:
Addressed all comments from @mellenburg and @branden and removed the obsolete AdminRouter bump.
Ready for review (and hopefully approval) once the CI turns all green again
